rfd/00TBD-webui-paginate-filter.md

```diff
@@ -0,0 +1,335 @@
+| App.Metadata.Name        | "colName"        |
+| App.Metadata.Description | "colDescription" |
+| App.Spec.PublicAddr      | "colPublicAddr"  |
+
+Given a query string, build an expression tree to create field conditions:
+```go
+// row must contain labels `env:prod` and `os:mac` and search value `searchValue`
+equals(env, "prod") && equals(os, "mac") && (contains(colName, "searchValue") || contains(colDescription, "searchValue") || contains(colPublicAddr, "searchValue"))
+```
+
+For each row, construct a `map[string]string` that contains all fields and values we want to test against our field conditions:
+```go
+Fields{
+  // Static Labels, already in map[string]string format
+  ...App.Metadata.labels,
+  // Dynamic labels, already in map[string]string format
+  ...App.Spec.CmdLabels
+  // Searcheable fields
+  "colName": App.Metadata.Name,
+  "colDescription": App.Metadata.Description,
+  "colPublicAddr": App.Spec.PublicAddr
+}
+```
+
+**Some things to consider:**
+
+Sometimes the UI will format a field to be displayed differently ie: if a server does not have a addr, we display `tunnel` to the user, a user may search for `tunnel`, expecting to get all tunnels. Or for app types, we display `cloud sql` instead of the actual value `gcp`.
+
+I'm not entirely sure how to handle this, other than to display actual values to user (tunnel column and display `gcp` instead of `cloud sql`). Or if filtering by tunnels is important, we can make `isTunnel` to be a type of filter a user can specify, and if specified returns servers that uses tunnels.
+
+<br/>
+
+### List of Unique Labels
+
+A user needs to be able to see a list of available labels as a dropdown, and so we need to create a new rpc and a new web api endpoint to retrieve them.
+
+```
+rpc GetResourceLabels(resourceType string) returns []Labels
+
+GET /webabpi/sites/:sites/<nodes|apps|databases|kubernetes>/labels
+```
+
+The new rpc will return a list of unique labels by going through a list of resources that is first ran through rbac checks (so users don't see labels to a resource they don't have access to).
+
+There is an edge case where each resource in a list, can have unique labels which can result in a large list, ie. with 30k items in list, one unique label will result in 30k+ labels, two unique labels will result in 60k+ labels and so on. I don't think listing one time labels are particularly useful to the user. We can only return labels that have at least 2 occurences by keeping track of how often labels appear as we process the list to dramatically cut down the length.
+
+There are two types of labels: `static` and `dynamic` ie:
+
+```
+ssh_service:
+  enabled: yes
+
+  # static
+  labels:
+    env: dev
+
+  # dynamic
+  commands:
+  - name: uptime
+    command: ["uptime", "-p"]
+    period: 1m0s
+```
+
+Some `dynamic` labels are time sensitive, as shown in the example. Such `dyanmic` labels with non zero `duration` will not be part of the list.
+
```

Comment:
While this particular example is something that makes a lot of sense to omit, some dynamic commands might change *very* infrequently and have a small number of possible values (e.g. if checking the version of installed software).  IMO its probably simpler to ignore the distinction.  Something like `uptime` probably won't have many duplicates with the same value anyway.  The system of preferentially showing values that appear more commonly will probably be sufficient to filter it out.